test/unit/StakingPool/calculatePremium.js

```diff
@@ -0,0 +1,480 @@
+    // calculate premiums
+    const basePrice = await calculateBasePrice(stakingPool, product, PRICE_CHANGE_PER_DAY);
+    const basePremium = amount.mul(basePrice).div(INITIAL_PRICE_DENOMINATOR);
+    const { surgePremium, surgePremiumSkipped } = await calculateSurgePremium(
+      stakingPool,
+      amount,
+      0 /* activeAllocations */,
+      totalCapacity,
+      this.config,
+    );
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    const calculatedPremium = await stakingPool.calculatePremium(
+      product,
+      buyCoverParams.period,
+      amount.div(NXM_PER_ALLOCATION_UNIT), // cover amount
+      0, // initialCapacityUsed
+      totalCapacity,
+      product.targetPrice, // targetPrice
+      timestamp, // current block timestamp
+    );
+    const expectedPremium = basePremium.add(surgePremium).sub(surgePremiumSkipped).div(periodsInYear);
+    expect(calculatedPremium.premium).to.be.equal(expectedPremium);
+    await cover.connect(coverBuyer).allocateCapacity(buyCoverParams, coverId, stakingPool.address);
+
+    // get active allocations
+    const activeAllocations = await stakingPool.getActiveAllocations(productId);
+    const totalActiveAllocations = activeAllocations.reduce(
+      (acc, allocation) => acc.add(allocation),
+      BigNumber.from(0),
+    );
+    expect(totalActiveAllocations).to.be.equal(totalCapacity);
+    expect(await cover.lastPremium()).to.be.equal(expectedPremium);
+  });
+
+  it('should should overflow uint32 tranche allocation when cover amount is too large', async function () {
+    const { stakingPool, cover } = this;
+    const [coverBuyer, staker] = this.accounts.members;
+    const amount = BigNumber.from(2).pow(96).sub(1);
+    const buyCoverParams = { ...buyCoverParamsTemplate, amount };
+
+    await stakingPool.connect(staker).depositTo([
+      {
+        tokenId: 0,
+        amount,
+        destination: staker.address,
+        trancheId: (await getCurrentTrancheId()) + 3,
+      },
+    ]);
+
+    await expect(
+      cover.connect(coverBuyer).allocateCapacity(buyCoverParams, coverId, stakingPool.address),
+    ).to.be.revertedWith("SafeCast: value doesn't fit in 32 bits");
+  });
+
+  it('should correctly calculate the premium', async function () {
+    const { stakingPool, cover } = this;
+    const { GLOBAL_CAPACITY_RATIO, PRICE_CHANGE_PER_DAY, INITIAL_PRICE_DENOMINATOR } = this.config;
+
+    const { totalCapacity } = await stakingPool.getActiveTrancheCapacities(
+      buyCoverParamsTemplate.productId,
+      GLOBAL_CAPACITY_RATIO,
+      coverProductTemplate.capacityReductionRatio,
+    );
+
+    // 1st cover buy
+    {
+      await moveDaysForward(183);
+
+      const product = await stakingPool.products(productId);
+      const basePrice = await calculateBasePrice(stakingPool, product, PRICE_CHANGE_PER_DAY);
+      expect(basePrice).to.be.equal(200);
+      const premiumPerYear = await stakingPool.calculatePremiumPerYear(
```

Comment:
What you are testing here is not the premium calculation but rather the fact that the staking pool returns the result of the function correctly to the cover contract. If the `calculatePremiumPerYear` would return a crazy value this test would pass without noticing.